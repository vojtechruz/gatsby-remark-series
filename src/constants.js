/**
 * Default series toc cache key.
 */
export const SERIES_TOC_CACHE_KEY = "__GATSBY_REMARK_SERIES_TOC";

/**
 * Default settings for the series toc.
 */
export const DefaultSettings = {
  field: "series"
};
