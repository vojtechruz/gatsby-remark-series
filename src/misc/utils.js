import kebabCase from "lodash.kebabcase";
import defaultsDeep from "lodash.defaultsdeep";

/**
 * Default template.
 * @param {TemplateContext} context The context to be used.
 * @returns {string} The default template.
 */
export const DefaultTemplate = context => {
  let body;
  let title;

  if (context.pluginOptions.render.useLandingPage === true) {
    body = "";
    title = `<span>This post is part of the <a href="${context.slug}">${
      context.name
    }</a> series</span>`;
  } else {
    const list = context.items.reduce((code, item) => {
      const inner =
        item.draft !== true
          ? `<a href="${item.slug}">${item.title}</a>`
          : item.title;

      return `${code}<li>${inner}</li>`.trim();
    }, "");

    body = `<ol>${list}</ol>`;
    title = `<div>Other posts in the ${context.name} series</div>`;
  }

  return `
<div class="series-table-of-content">
  ${title}
  ${body}
</div>
  `.trim();
};

/**
 * Default options for the series.
 */
export const DefaultOptions = {
  resolvers: {
    slug: markdownNode => markdownNode.frontmatter.slug,
    date: markdownNode => markdownNode.frontmatter.date,
    draft: markdownNode => markdownNode.frontmatter.draft,
    order: markdownNode => markdownNode.frontmatter.order,
    series: markdownNode => markdownNode.frontmatter.series,
    toSlug: series => kebabCase(series)
  },
  render: {
    mode: "inline",
    placeholder: "bottom",
    template: DefaultTemplate,
    useLandingPage: false
  }
};

/**
 * Checks the type matches.
 * @param {string} type The expected type.
 * @param {string} name The name of the property.
 * @param {*} value The value of the property.
 */
function assertType(type, name, value) {
  if (typeof value !== type) {
    throw new Error(`${name} must be of type ${type}`);
  }
}

/**
 * Resolves the plug-in options to be used.
 * @param {PluginOptions} pluginOptions The user defined options.
 * @returns {PluginOptions} The resolved options.
 */
export function resolveOptions(pluginOptions) {
  const options = defaultsDeep({}, pluginOptions, DefaultOptions);

  assertType("string", "mode", options.render.mode);
  assertType("string", "placeholder", options.render.placeholder);
  assertType("function", "template", options.render.template);
  assertType("boolean", "useLandingPage", options.render.useLandingPage);

  options.render.mode = options.render.mode.toLowerCase();
  options.render.placeholder = options.render.placeholder.toLowerCase();

  assertType("function", "slug", options.resolvers.slug);
  assertType("function", "date", options.resolvers.date);
  assertType("function", "draft", options.resolvers.draft);
  assertType("function", "order", options.resolvers.order);
  assertType("function", "series", options.resolvers.series);
  assertType("function", "toSlug", options.resolvers.toSlug);

  if (options.render.useLandingPage === true) {
    assertType(
      "string",
      "landingPageComponent",
      options.render.landingPageComponent
    );

    if (options.render.landingPagePathPrefix != null) {
      assertType(
        "string",
        "landingPagePathPrefix",
        options.render.landingPagePathPrefix
      );
    }
  }

  return options;
}

/**
 * Resolves the series path (slug).
 * @param {string} name The name of the series.
 * @param {string} [pathPrefix] The global path prefix to be used.
 * @param {string} [pluginPathPrefix] The plugin defined path prefix to be used.
 * @returns {string} The resolved series path (slug).
 */
export function resolveSeriesPath(name, pathPrefix, pluginPathPrefix) {
  const path = [];

  assertType("string", "name", name);

  if (typeof pathPrefix === "string") {
    const cleanPathPrefix = pathPrefix.replace(/^\/|\/$/g, "");
    if (cleanPathPrefix.trim().length > 0) {
      path.push(cleanPathPrefix);
    }
  }

  if (typeof pluginPathPrefix === "string") {
    const cleanpluginPathPrefix = pluginPathPrefix.replace(/^\/|\/$/g, "");
    if (cleanpluginPathPrefix.trim().length > 0) {
      path.push(cleanpluginPathPrefix);
    }
  }

  path.push(name.replace(/^\/|\/$/g, ""));

  return `/${path.join("/")}`;
}

/**
 * Sorts the items in a series.
 * Items are sorted by order, falling back to date then title.
 * @param {SeriesItem} left The left item to be sorted.
 * @param {SeriesItem} right The right item to be sorted.
 * @returns {number} The sorted resolution.
 */
export function sortItems(left, right) {
  const orderLeft = left.order !== null ? Number(left.order) : NaN;
  const orderRight = right.order !== null ? Number(right.order) : NaN;

  if (isNaN(orderLeft) || isNaN(orderRight)) {
    if (!isNaN(orderLeft)) return -1;
    if (!isNaN(orderRight)) return 1;
  } else if (orderLeft !== orderRight) {
    return orderLeft - orderRight;
  }

  const dateLeft =
    left.date !== null ? new Date(left.date) : new Date(undefined);
  const dateRight =
    right.date !== null ? new Date(right.date) : new Date(undefined);

  if (isNaN(dateLeft.getTime()) || isNaN(dateRight.getTime())) {
    if (!isNaN(dateLeft.getTime())) return -1;
    if (!isNaN(dateRight.getTime())) return 1;
  } else if (dateLeft.getTime() !== dateRight.getTime()) {
    return dateLeft.getTime() - dateRight.getTime();
  }

  if (left.title == null || right.title == null) {
    if (left.title != null) return -1;
    if (right.title != null) return 1;
    return 0;
  }

  return left.title.localeCompare(right.title);
}
