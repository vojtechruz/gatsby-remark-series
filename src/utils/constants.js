/**
 * Default series toc cache key.
 */
export const SERIES_TOC_CACHE_KEY = "__GATSBY_REMARK_SERIES_TOC__";

/**
 * Default settings for the series toc.
 */
export const DefaultSettings = {};
