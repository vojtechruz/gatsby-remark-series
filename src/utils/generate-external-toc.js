export default async (context, settings) => {
  const { markdownNode, markdownAST } = context;
};
