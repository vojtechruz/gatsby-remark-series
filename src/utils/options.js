import defaultsDeep from "lodash.defaultsdeep";

import { DefaultSettings } from "./constants";

/**
 * Resolves the plug-in options to be used.
 * @param {PluginOptions} options The user defined options.
 * @returns {PluginOptions} The resolved options.
 */
export function resolveOptions(options) {
  return defaultsDeep({}, options, DefaultSettings);
}

/**
 * Resolves the series path (slug).
 * @param {string} name The name of the series.
 * @param {string} [pathPrefix] The global path prefix to be used.
 * @param {string} [userPathPrefix] The user defined path prefix to be used.
 * @returns {string} The resolved series path (slug).
 */
export function resolveSeriesPath(name, pathPrefix, userPathPrefix) {
  const path = [];

  if (typeof pathPrefix === "string" && pathPrefix.trim().length > 0) {
    path.push(pathPrefix);
  }

  if (typeof userPathPrefix === "string" && userPathPrefix.trim().length > 0) {
    path.push(userPathPrefix);
  }

  path.push(name);

  return path.join("/");
}
