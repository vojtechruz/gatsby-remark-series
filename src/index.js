import { external, inline } from "./misc/table-of-content";
import { resolveOptions } from "./misc/utils";

/**
 * Handles the markdown AST.
 * @param {RemarkPluginContext} context the remark plugin context.
 * @param {PluginOptions} pluginOptions the options of the plugin.
 * @returns {*} the markdown ast.
 */
export default (context, pluginOptions) => {
  const options = resolveOptions(pluginOptions);

  return options.render.mode === "external"
    ? external(context, options)
    : inline(context, options);
};
