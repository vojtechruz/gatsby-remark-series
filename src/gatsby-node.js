import { resolveSeriesPath } from "./index/options";

/** @inheritdoc */
export async function createPages(
  { graphql, pathPrefix, actions: { createPage } },
  options
) {
  if (options.render.mode !== "external") return;

  const result = await graphql(
    `
      {
        series: allMarkdownRemark {
          edges {
            node {
              frontmatter {
                date
                title
                ${options.draftField}
                ${options.orderField}
                ${options.seriesField}
              }
            }
          }
        }
      }
    `
  );

  const series = result.data.all.edges.reduce((map, node) => {
    const name = node.frontmatter.series;

    if (name == null) return map;

    map[name] = map[name] || [];
    map[name].push({ ...node.fields.slug, ...node.frontmatter });

    return map;
  }, {});

  Object.keys(series).map(key => {
    createPage({
      path: resolveSeriesPath(key, pathPrefix, options.pathPrefix),
      component: options.externalLayout,
      context: {
        items: series[key]
      }
    });
  });
}
