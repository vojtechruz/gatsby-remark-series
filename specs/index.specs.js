import { readFile } from "fs";
import { promisify } from "util";

import Remark from "remark";
import grayMatter from "gray-matter";
import defaultsDeep from "lodash.defaultsdeep";

import plugin from "../src/index";
import { DefaultOptions } from "../src/utils";

describe("gatsby-remark-series/remark", () => {
  const reporter = {
    panic: message => {
      throw new Error(message);
    }
  };

  const remark = new Remark().data("settings", {
    commonmark: true,
    footnotes: true,
    pedantic: true
  });

  const createContentDigest = () => {
    return Math.random()
      .toString(36)
      .substring(7);
  };

  const getNodes = () => {
    return [
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title 2",
          date: "2019-01-01",
          series: "My Title Series"
        }
      },
      { internal: { type: "other" } },
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title 3",
          date: "2019-02-01",
          series: "My Title Series"
        }
      },
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title",
          date: "2019-01-01",
          series: "Other Series"
        }
      },
      { internal: { type: "other" } }
    ];
  };

  const readFilePromise = promisify(readFile);

  const getSeriesMarkdown = async (file, options, pathPrefix) => {
    const seriesMarkdown = await readFilePromise(file, "utf-8");
    const markdownAST = remark.parse(seriesMarkdown.trim());
    const countPrePlugin = markdownAST.children.length;

    const postMarkdownAST = plugin(
      {
        createContentDigest,
        getNodes,
        markdownAST,
        markdownNode: { frontmatter: grayMatter(seriesMarkdown.trim()).data },
        pathPrefix,
        reporter
      },
      options
    );

    return [postMarkdownAST, countPrePlugin];
  };

  test("returns the exact AST when no series is detected", async () => {
    const noSeriesMarkdown = await readFilePromise(
      "./specs/files/no-series.md",
      "utf-8"
    );

    const markdownAST = remark.parse(noSeriesMarkdown.trim());
    const countPrePlugin = markdownAST.children.length;

    const postMarkdownAST = plugin(
      {
        createContentDigest,
        markdownAST,
        markdownNode: { frontmatter: grayMatter(noSeriesMarkdown.trim()).data },
        reporter
      },
      DefaultOptions
    );

    expect(countPrePlugin).toBe(postMarkdownAST.children.length);
    expect(postMarkdownAST).toMatchSnapshot();
  });

  test("returns a modified version of the AST when series is detected", async () => {
    const [postMarkdownAST, countPrePlugin] = await getSeriesMarkdown(
      "./specs/files/series.md",
      DefaultOptions
    );

    expect(countPrePlugin).not.toBe(postMarkdownAST.children.length);
    expect(postMarkdownAST).toMatchSnapshot();
  });

  test("replaces children's value when dynamic placeholder is used", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series-dynamic.md",
      defaultsDeep({ render: { placeholder: "<!-- TOC -->" } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      "series-table-of-content"
    );
  });

  test("adds a children on top of the AST's children when top placeholder is used", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep({ render: { placeholder: "top" } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[0].type).toBe("html");
    expect(postMarkdownAST.children[0].value).toContain(
      "series-table-of-content"
    );
  });

  test("adds a children on bottom of the AST's children when bottom placeholder is used", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep({ render: { placeholder: "bottom" } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      "series-table-of-content"
    );
  });

  test("renders inline toc references", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep({ render: { useLandingPage: false } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      "Other posts in the My Title Series series"
    );
  });

  test("renders external toc references", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep(
        {
          render: {
            useLandingPage: true,
            landingPageComponent: "path"
          }
        },
        DefaultOptions
      )
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/my-title-series">My Title Series</a>'
    );
  });

  test("uses the pluginPathPrefix in external toc", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep(
        {
          render: {
            useLandingPage: true,
            landingPageComponent: "path",
            landingPagePathPrefix: "series"
          }
        },
        DefaultOptions
      )
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/series/my-title-series">My Title Series</a>'
    );
  });

  test("uses the pathPrefix in external toc", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      "./specs/files/series.md",
      defaultsDeep(
        {
          render: {
            useLandingPage: true,
            landingPageComponent: "path",
            landingPagePathPrefix: "series"
          }
        },
        DefaultOptions
      ),
      "blog"
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/blog/series/my-title-series">My Title Series</a>'
    );
  });
});
