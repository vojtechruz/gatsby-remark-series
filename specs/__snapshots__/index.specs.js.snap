// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-remark-series/remark passes options to generate-toc 1`] = `
Object {
  "children": Array [
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-series/remark resolves options before sending to generate-toc 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`gatsby-remark-series/remark resolves options before sending to generate-toc 2`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
